Refactor voice to separate it from sex

ZigTag opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
Refactors the entire voice system! Also adds a selector that people can use to change their voice!

This originally was a PR I made in an AGPL fork however I've decided to bring it up here for everyone to enjoy

Closes #38461 

PR Hotfix #44468

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
Sometimes people don't have the same voice as their body, this PR fixes that! This also removes the association between sex and voice entirely, therefore you can have an infinite amount of voices!

## Technical details
<!-- Summary of code changes for easier review. -->
- Adds a new database item named Voice, stores the ProtoId<EmoteSoundsPrototype> as a string
- Voice selector in the character creator
  - This displays all the ProtoIds listed in the SpeciesPrototype voices set
- Moved sex association from VocalComponent to SpeciesPrototype
  - It is hard coded to read Male, Female, Unisex, in that order and all 3 must be defined. It is a fixed length array. This is for performance

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->
New UI:
<img width="683" height="243" alt="image" src="https://github.com/user-attachments/assets/c6edfa3a-6cab-44b9-abe1-1d355c40b966" />

Voice Selector on unsexed
<img width="698" height="238" alt="image" src="https://github.com/user-attachments/assets/6f37796d-ecab-4a8f-812f-99142ad89188" />

Voice Selector on sexed
<img width="683" height="231" alt="image" src="https://github.com/user-attachments/assets/8f14c5d3-faa7-4993-8c80-0b46a848d78b" />

Example of Female sex with Masculine voice
https://github.com/user-attachments/assets/321ee630-15fb-4c68-ac7d-ed5acaa1d6b1


## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [x] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->
- SexChangedEvent has been removed. The only existing usage was for voices, which is now separated from sex. Anything referring to voices should use the new VoiceChangedEvent, otherwise there is no replacement event for sex.
- The field which used to be `sounds` is now removed from VocalComponent and now lives in SpeciesPrototype. You are encouraged to pick a specific voice from the bank and set it directly using the emoteSounds property on VocalComponent for non-round-start species (including non-humanoid mobs).
- Things like DB conversion should already be handled
  - From what I could tell it works fine, but please make sure when testing
- There is a new field on SpeciesPrototype which is called Voices, this is a list of user selectable voices in the UI, and the name displayed is controlled via the name field in the EmoteSoundsPrototype.
  - Currently, the migration steps are as follows.
     1. Consolidate all vocal `sounds` to a single `emoteSounds` entry, this is considered "the default" (like for urists or sheep)
         - The regex `Vocal[\r\n]+.*sounds:` worked well for me for finding all these.
         - Make sure to copy the sexed entries to the species prototypes on all species under the defaultSoundsBySex property, most likely you just need to remove the `Sex:` part of them and make it as a list in the order of Male, Female, and Unsexed.
     2. Add the voice category (the list of EmoteSoundsPrototypes you would like to display on the UI) and move the sounds (now called defaultSoundsBySex) on all species to the SpeciesPrototype (usually found in ./Resources/Prototypes/Species/*)
     3. Add Localization for emote sounds prototypes:
         - On unsexed banks, you can use `humanoid-profile-editor-voice-neutral` as the name field
         - On sexed banks you can use `humanoid-profile-editor-voice-masculine` for Male banks and `humanoid-profile-editor-voice-feminine` for female banks
<img width="563" height="374" alt="image" src="https://github.com/user-attachments/assets/e85fab76-1d79-4f06-afe5-bc57aaaa7aa0" />
<img width="417" height="322" alt="image" src="https://github.com/user-attachments/assets/207577d6-d476-4256-974e-c3029b9c406a" />

And for mobs
<img width="605" height="366" alt="image" src="https://github.com/user-attachments/assets/633b31b7-d715-4377-a906-f8e2888d975f" />

This was done because Sex doesn't actually exist on non-humanoids, and they only used the Sex.Unsexed entries, as that's the default sex. Its significantly more ergonomic to set the bank directly. 

Here's an example of what the FTLs on emoteSoundsPrototype(s) look like:
<img width="578" height="464" alt="image" src="https://github.com/user-attachments/assets/5faffe32-4e0a-4124-92b4-8783a335457d" />

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
:cl:
- add: your voice is now selectable
- tweak: game now supports having multiple voice sets per species

ZigTag committed:
feat: separate sex from voice

MozarteanChaos commented:
This would close #38461 if merged.

ZigTag commented:
> You should probably change the changelog to something else, but I think this is based c:

You're right, should be fixed now

beck-thompson commented on Content.Shared/Preferences/HumanoidCharacterProfile.cs:
```diff
@@ -80,6 +80,9 @@ public sealed partial class HumanoidCharacterProfile : ICharacterProfile
         public Sex Sex { get; private set; } = Sex.Male;
 
+        [DataField]
+        public Sex? PreferredVoice { get; set; }
```
The voice shouldn't be using Sex, if your going to decouple it go all the way! If I'm misunderstanding let me know

<img width="789" height="388" alt="Image" src="https://github.com/user-attachments/assets/a14d2762-c096-42ed-a0f0-518c09a675b0" />

ZigTag replied:
If you prefer switching to it's own enum should we be checking the list of setable voices instead? My only concession is that it requires SpeciesProto -> MobProto -> VocalComponent

beck-thompson commented on Content.Shared/Humanoid/SharedHumanoidAppearanceSystem.cs:
```diff
@@ -366,10 +367,44 @@ public void SetSex(EntityUid uid, Sex sex, bool sync = true, HumanoidAppearanceC
+        // Casts unrestricted selector into restricted sex settings
+        List<Sex> sexes = new();
+
+        if (_proto.TryIndex(humanoid.Species, out var speciesProto))
```
Use `Index`, `TryIndex` should never fail, and if it does you want it to error

ZigTag replied:
Fair enough!

beck-thompson commented on Content.Shared/Humanoid/SharedHumanoidAppearanceSystem.cs:
```diff
@@ -366,10 +367,44 @@ public void SetSex(EntityUid uid, Sex sex, bool sync = true, HumanoidAppearanceC
+            Dirty(uid, humanoid);
+    }
+
+    // TheDen - Add Voice
```
remove these!

ZigTag replied:
Ah shoot I thought I stripped all those, good catch!

beck-thompson commented on Content.Shared/Humanoid/HumanoidAppearanceComponent.cs:
```diff
@@ -26,6 +26,9 @@ public sealed partial class HumanoidAppearanceComponent : Component
     // Couldn't these be somewhere else?
 
+    [DataField, AutoNetworkedField]
+    public Sex PreferredVoice;
```
Add comment

beck-thompson reviewed (dismissed):
woke station 14 real

ZigTag committed:
Merge upstream/master into feat/voice-selector

ZigTag committed:
wip: tried to make it use voice collection directly

ZigTag committed:
fix: yaml

ZigTag committed:
fix: migrations

ZigTag committed:
fix: ensure valid now validates correctly

ZigTag committed:
fix: unisex mobs wouldn't have voice

ZigTag committed:
chore: fix RT version

ZigTag committed:
refactor: clean up code into subfunctions

ZigTag committed:
refactor: swap name and protoId in voice

also rename sounds to voice

ZigTag committed:
chore: add comment

ZigTag committed:
fix: fallback sounds for non-humanoid characters

ZigTag committed:
fix: tests

ZigTag committed:
fix: tests

sowelipililimute reviewed (dismissed):
some nits while i work on a more comprehensive review

ZigTag committed:
Merge branch 'master' into feat/voice-selector

ZigTag committed:
refactor: change requests

ZigTag committed:
Merge branch 'master' into feat/voice-selector

ZigTag committed:
refactor: convert sheepers

ZigTag committed:
refactor: move Voice to SpeciesPrototype

ZigTag committed:
chore: add tests

ZigTag committed:
fix: test IoC error

ZigTag committed:
refactor: use null coalesing in ConvertProfiles

ZigTag committed:
refactor: move vocal sex association to speciesPrototype

ZigTag committed:
style: remove space from new()

ZigTag committed:
refactor: use name from emoteSoundsPrototype for UI

ZigTag committed:
refactor: no more null Voice on profile

ZigTag committed:
refactor: make DefaultVoice a static readonly property

ZigTag committed:
fix: add default voice to HumanoidProfileComponent

ZigTag committed:
chore: remove repeated vocal component from species yaml

It's inherited from BaseSpeciesMob

ZigTag committed:
style: correct comments for refactor

ZigTag committed:
fix: add missing locale for FemaleVulpkanin

ZigTag committed:
refactor: simplify logic based off surounding code

ZigTag committed:
test: add test to make sure that sexes are properly enumerated for the voices

ZigTag committed:
fix: add back urist sounds

ZigTag committed:
style: clarify comments again

ZigTag committed:
style: last thing im changing before I sleep

ZigTag committed:
refactor: battling programming semantics

ZigTag committed:
refactor: Name on EmoteSoundsPrototype.cs to VoiceSelectorName for clarity

ZigTag committed:
revert: isDirty change

AJCM-git committed:
Merge branch 'master' into pr/5

AJCM-git committed:
Fix HumanoidProfileTest.EnsureValidLoading.
    The profile's voice wasnt being updated, thus the voiceChanged event was never raised and that meant that the voice component had old EmoteSounds

AJCM-git committed:
oops

ZigTag committed:
refactor: do the easy change requests

ZigTag committed:
refactor: use an array for indexing sex conversion for performance

ZigTag committed:
Merge branch 'master' into 'feat/voice-selector'

AJCM-git committed:
Merge branch 'master' into pr/40593

AJCM-git reviewed (approved):
lgtm

AJCM-git committed:
fix master

AJCM-git committed:
mistakes were made

ZigTag committed:
chore: redo migrations
